src/resolve.ts

```diff
@@ -59,11 +76,141 @@ async function loadCache(cacheFilePath: string): Promise<CacheSchemaV3> {
+      // See https://github.com/yarnpkg/yarn/blob/953c8b6a20e360b097625d64189e6e56ed813e0f/src/resolvers/exotics/github-resolver.js#L6
+      packagePattern.startsWith('github:') ||
+      /^[^:@%/\s.-][^:@%/\s]*[/][^:@\s/%]+(?:#.*)?$/.test(packagePattern) ||
+      // See https://github.com/yarnpkg/yarn/blob/953c8b6a20e360b097625d64189e6e56ed813e0f/src/resolvers/exotics/gitlab-resolver.js#L6
+      packagePattern.startsWith('gitlab:') ||
+      // See https://github.com/yarnpkg/yarn/blob/953c8b6a20e360b097625d64189e6e56ed813e0f/src/resolvers/exotics/bitbucket-resolver.js#L6
+      packagePattern.startsWith('bitbucket:') ||
+      // See https://github.com/yarnpkg/yarn/blob/953c8b6a20e360b097625d64189e6e56ed813e0f/src/resolvers/exotics/gist-resolver.js#L26
+      packagePattern.startsWith('gist:') ||
+      // See https://github.com/yarnpkg/yarn/blob/953c8b6a20e360b097625d64189e6e56ed813e0f/src/resolvers/exotics/git-resolver.js#L19
+      /^git:|^git\+.+:|^ssh:|^https?:.+\.git$|^https?:.+\.git#.+/.test(packagePattern)) {
+    return true
+  } else {
+    // See https://github.com/yarnpkg/yarn/blob/953c8b6a20e360b097625d64189e6e56ed813e0f/src/resolvers/exotics/git-resolver.js#L19
+    const {hostname, path} = url.parse(packagePattern);
+    if (hostname && path && ['github.com', 'gitlab.com', 'bitbucket.com', 'bitbucket.org'].indexOf(hostname) >= 0) {
+      return path.split('/').filter((p): boolean => !!p).length === 2;
+    }
+  }
+}
+
+function resolvePackageNameFromRegistryYarnPackagePattern(packagePattern: string): string {
+  // See https://github.com/yarnpkg/yarn/blob/953c8b6a20e360b097625d64189e6e56ed813e0f/src/resolvers/exotics/registry-resolver.js#L12
+  const match = packagePattern.match(/^(\S+):(@?.*?)(@(.*?)|)$/);
+  if (match) {
+    return match[2];
+  } else {
+    throw new Error(`Failed to resolve yarn package pattern ${packagePattern}, unrecognized format`)
+  }
+}
+
+function resolveNpmPackagesFromYarnLockDependencies(yarnLockDependencies: YarnLockDependency[]): PackageWithId[] {
+  const packages: PackageWithId[] = yarnLockDependencies
+    .reduce(function(filterMappedDependencies: PackageWithId[], {packagePattern, version}) {
+      if (isNonRegistryYarnPackagePattern(packagePattern)) {
+        return filterMappedDependencies;
+      }
+
+      let packageName
+
+      if (packagePattern.startsWith('npm:') || packagePattern.startsWith('yarn:')) {
+        packageName = resolvePackageNameFromRegistryYarnPackagePattern(packagePattern)
+      } else {
+        // Package pattern not yet recognized, continue with parsing logic from
+        // https://github.com/yarnpkg/yarn/blob/953c8b6a20e360b097625d64189e6e56ed813e0f/src/package-request.js#L99
+        const {name: namePart, range: rangePart} = normalizeYarnPackagePattern(packagePattern)
+
+        if (isNonRegistryYarnPackagePattern(rangePart)) {
+          return filterMappedDependencies;
```

Comment:
..and get rid of ;s (there are a few here and there)

Reply:
361196e31bcb104a8e1885630427cdffa257b543
Removed semicolons. I was supposed go through them but it slipped my mind :/
It actually felt weird to write this code without prettier automatically styling everything consistently. If you're open to accepting a PR where prettier is taken into use, I might do one 😄

Reply:
have not really used it, so now would be a good time to try :-)